src/main/java/com/nutomic/syncthingandroid/preferences/WifiSsidPreference.java

```diff
@@ -0,0 +1,249 @@
+        OnPreferenceChangeListener listener = getOnPreferenceChangeListener();
+        if (listener != null) {
+            listener.onPreferenceChange(this, ssids);
+        }
+    }
+
+    /**
+     * Adapter between the available and selected Wifis in the preference.
+     */
+    class WifiListAdapter extends BaseAdapter {
+
+        /**
+         * @return either the number of configured networks, or 1 if information unavailable
+         */
+        @Override
+        public int getCount() {
+            return mConfiguredNetworks != null ? mConfiguredNetworks.length : 1;
+        }
+
+        @Override
+        public Object getItem(int position) {
+            return mConfiguredNetworks != null ? mConfiguredNetworks[position] : null;
+        }
+
+        @Override
+        public long getItemId(int position) {
+            return position;
+        }
+
+        @Override
+        public View getView(int position, View convertView, ViewGroup parent) {
+            if (mConfiguredNetworks != null) {
+                WifiConfiguration wifi = mConfiguredNetworks[position];
+                boolean viewCreated = false;
+
+                // Create view only if new or reuse of provided view is not possible
+                if (convertView == null
+                        || !(convertView.findViewById(R.id.wifiName) instanceof CheckBox)) {
+                    LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+                    convertView = inflater.inflate(R.layout.preference_wifi_ssid_row, parent, false);
+                    viewCreated = true;
+                }
+
+                // SSIDs are formatted "name" if they represent an UTF-8 string,
+                // or as 01234567890ABCDEF if hex, but displaying the '"' doesn't look good,
+                // so strip them
+                String displayName = wifi.SSID;
+                if (displayName != null && displayName.length() > 0 && displayName.charAt(0) == '"') {
+                    // If th SSID starts with '"', it is a string name, otherwise a hex string
+                    if (displayName.length() >= 2) {
+                        // actually, this should always be true, otherwise it's a violation of the
+                        // contract of how SSID names are formatted by WifiManager
+                        displayName = displayName.substring(1, displayName.length() - 1);
+                    }
+                }
+
+                CheckBox wifiView = (CheckBox) convertView.findViewById(R.id.wifiName);
+                wifiView.setTag(wifi.SSID); // the SSID this possibly recylced view is currently representing6
+                wifiView.setText(displayName);
+                wifiView.setChecked(mSelectedSsids.contains(wifi.SSID));
+
+                if (viewCreated) {
+                    wifiView.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
```

Comment:
It's easier if you set a new listener every time, and just access final variables from the surrounding function. Leave something to do for the GC :D